src/webgpu/web_platform/canvas/readbackFromWebGPUCanvas.spec.ts

```diff
@@ -70,29 +72,89 @@ async function initCanvasContent(
-    format: 'bgra8unorm',
-    usage: GPUTextureUsage.COPY_DST | GPUTextureUsage.COPY_SRC,
+    format,
+    usage: GPUTextureUsage.COPY_SRC | GPUTextureUsage.RENDER_ATTACHMENT,
   });
 
-  const rows = 2;
-  const bytesPerRow = 256;
-  const buffer = t.device.createBuffer({
-    mappedAtCreation: true,
-    size: rows * bytesPerRow,
-    usage: GPUBufferUsage.COPY_SRC,
+  const pipeline = t.device.createRenderPipeline({
+    vertex: {
+      module: t.device.createShaderModule({
+        code: `
+struct VertexOutput {
+  @builtin(position) Position : vec4<f32>;
+};
+
+@stage(vertex)
+fn main(@builtin(vertex_index) VertexIndex : u32) -> VertexOutput {
+  var pos = array<vec2<f32>, 6>(
+      vec2<f32>( 1.0,  1.0),
+      vec2<f32>( 1.0, -1.0),
+      vec2<f32>(-1.0, -1.0),
+      vec2<f32>( 1.0,  1.0),
+      vec2<f32>(-1.0, -1.0),
+      vec2<f32>(-1.0,  1.0));
+
+  var output : VertexOutput;
+  output.Position = vec4<f32>(pos[VertexIndex], 0.0, 1.0);
+  return output;
+}
+            `,
+      }),
+      entryPoint: 'main',
+    },
+    fragment: {
+      module: t.device.createShaderModule({
+        code: `
+@group(0) @binding(0) var mySampler: sampler;
+@group(0) @binding(1) var myTexture: texture_2d<f32>;
+
+@stage(fragment)
+fn main(@builtin(position) fragcoord: vec4<f32>) -> @location(0) vec4<f32> {
+  var coord = vec2<u32>(floor(fragcoord.xy));
+  var color = vec4<f32>(0.0, 0.0, 0.0, 1.0);
+  if (coord.x == 0u) {
+    if (coord.y == 0u) {
+      color.b = 1.0;
+    } else {
+      color.r = 1.0;
+    }
+  } else {
+    if (coord.y == 0u) {
+      color.g = 1.0;
+    } else {
+      color.r = 1.0;
+      color.g = 1.0;
+    }
+  }
+  return color;
+}
+            `,
+      }),
+      entryPoint: 'main',
+      targets: [{ format }],
+    },
+    primitive: {
+      topology: 'triangle-list',
+    },
   });
-  const mapping = buffer.getMappedRange();
-  const data = new Uint8Array(mapping);
-  data.set(new Uint8Array([0xff, 0x00, 0x00, 0xff]), 0); // blue
-  data.set(new Uint8Array([0x00, 0xff, 0x00, 0xff]), 4); // green
-  data.set(new Uint8Array([0x00, 0x00, 0xff, 0xff]), 256 + 0); // red
-  data.set(new Uint8Array([0x00, 0xff, 0xff, 0xff]), 256 + 4); // yellow
-  buffer.unmap();
-
-  const texture = ctx.getCurrentTexture();
-  const encoder = t.device.createCommandEncoder();
-  encoder.copyBufferToTexture({ buffer, bytesPerRow }, { texture }, [2, 2, 1]);
-  t.device.queue.submit([encoder.finish()]);
+
+  const renderPassDescriptor: GPURenderPassDescriptor = {
+    colorAttachments: [
+      {
+        view: ctx.getCurrentTexture().createView(),
+
+        loadValue: { r: 0.0, g: 0.0, b: 0.0, a: 1.0 },
+        storeOp: 'store',
+      },
+    ],
+  };
+
+  const commandEncoder = t.device.createCommandEncoder();
+  const passEncoder = commandEncoder.beginRenderPass(renderPassDescriptor);
+  passEncoder.setPipeline(pipeline);
+  passEncoder.draw(6, 1, 0, 0);
+  passEncoder.endPass();
+  t.device.queue.submit([commandEncoder.finish()]);
```

Comment:
I'm very sorry, I definitely didn't think this suggestion through. I was imagining using just the loadValue, no draw call. Of course this isn't possible because you want to draw different colors to each of the 4 pixels.

Using a draw call is obviously a lot more complex than using a copy even though the copy gets slightly complex with different formats. Simplicity would be nice here, but I suppose the draw call is still nicer since it will generalize to all formats.

An alternative would be to create a 1x1 temporary render target `temp` and do 4 empty render passes and 4 copyTextureToTexture commands:
- add local closure `const clearOnePixel = () => { ... }` that does:
    - beginRenderPass (temp, clear to a color, store), end pass
    - copyTextureToTexture(temp into one pixel of canvas)
- call `clearOnePixel` once for each of the 4 pixels

WDYT of this approach?